src/resolve/resolved-dns-transaction.c

```diff
@@ -1979,6 +1990,213 @@ static int mdns_make_dummy_packet(DnsTransaction *t, DnsPacket **ret_packet, Set
         return add_known_answers;
 }
 
+static size_t dns_transaction_curl_header_callback(void *contents, size_t size, size_t nmemb, void *userdata) {
+        _cleanup_free_ char *content_header = NULL;
+        DnsTransaction *t = ASSERT_PTR(userdata);
+        size_t sz = size * nmemb;
+        CURLcode code;
+        long status;
+        int r;
+
+        assert(contents);
+
+        code = curl_easy_getinfo(t->curl, CURLINFO_RESPONSE_CODE, &status);
+        if (code != CURLE_OK) {
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to retrieve response code: %s", curl_easy_strerror(code));
+        }
+
+        if (status >= 200 && status <= 299){
+                r = curl_header_strdup(contents, sz, "Content-Type:", &content_header);
+                if (r < 0) {
+                        log_oom();
+                        return 0;
+                }
+                if (r > 0){
+                        t->valid_dns_message = true;
+                        return sz;
+                }
+        }
+
+        return sz;
+}
+
+static size_t dns_transaction_curl_write_callback(void *contents, size_t size, size_t nmemb, void *userdata) {
+        DnsTransaction *t = ASSERT_PTR(userdata);
+        size_t sz = size * nmemb;
+        int r;
+
+        if (!GREEDY_REALLOC(t->payload, t->payload_size + sz)) {
+                r = log_oom();
+                goto fail;
+        }
+
+        t->payload = memdup(contents, sz);
```

Comment:
you leak `t->payload` here, i.e. the memory you allocated via GREEDY_REALLOC() is lost here where you allocate another memory block

Reply:
Thanks @poettering, I got rid of GREEDY_REALLOC() in favor of memdup(). If that is not right please re-open it.